src/cpptools.ts

```diff
@@ -37,52 +41,116 @@ interface TargetDefaults {
   defines: string[];
 }
 
-function parseCppStandard(std: string): StandardVersion|null {
+function parseCppStandard(std: string, can_use_gnu: boolean): StandardVersion|null {
+  const is_gnu = can_use_gnu && std.startsWith('gnu');
   if (std.endsWith('++2a') || std.endsWith('++20') || std.endsWith('++latest')) {
-    return 'c++20';
+    return is_gnu ? 'gnu++20' : 'c++20';
   } else if (std.endsWith('++17') || std.endsWith('++1z')) {
-    return 'c++17';
+    return is_gnu ? 'gnu++17' : 'c++17';
   } else if (std.endsWith('++14') || std.endsWith('++1y')) {
-    return 'c++14';
+    return is_gnu ? 'gnu++14' : 'c++14';
   } else if (std.endsWith('++11') || std.endsWith('++0x')) {
-    return 'c++11';
+    return is_gnu ? 'gnu++11' : 'c++11';
   } else if (std.endsWith('++03')) {
-    return 'c++03';
+    return is_gnu ? 'gnu++03' : 'c++03';
   } else if (std.endsWith('++98')) {
-    return 'c++98';
+    return is_gnu ? 'gnu++98' : 'c++98';
   } else {
     return null;
```

Comment:
Since `target` is expected to be lower-cased, a comment should be added to this effect.